File: operator/internal/controller/plugins/langchain.go
package plugins

import (
	"fmt"

	"github.com/example/agent-observability-operator/operator/internal/controller/plugins/common"
)

// LangChainPlugin implements InstrumentationPlugin for LangChain.
//
// LangChain does NOT support auto-detection because the official OTel instrumentor
// uses an all-or-nothing approach that instruments the entire LangChain ecosystem
// at once. This makes it unsafe to use auto-detection when apps may partially
// instrument LangChain components.
type LangChainPlugin struct{}

func (p *LangChainPlugin) Name() string {
	return "langchain"
}

func (p *LangChainPlugin) SupportsAutoDetection() bool {
	return false
}

// ============ Type Checking ============

func (p *LangChainPlugin) IsTrue(value interface{}) bool {
	return common.IsInterfaceTrue(value)
}

func (p *LangChainPlugin) IsFalse(value interface{}) bool {
	return common.IsInterfaceFalse(value)
}

func (p *LangChainPlugin) IsAuto(value interface{}) bool {
	return common.IsInterfaceAuto(value)
}

// ============ Field Resolution ============

func (p *LangChainPlugin) ResolveField(value interface{}, defaultBool bool) interface{} {
	return common.ResolveInterfaceField(value, defaultBool)
}

func (p *LangChainPlugin) ValueToString(value interface{}) string {
	return common.InterfaceToString(value)
}

// ============ Validation ============

func (p *LangChainPlugin) Validate(value interface{}) error {
	if p.IsAuto(value) {
		return fmt.Errorf(
			"langchain: auto is not supported. "+
				"The LangChain instrumentation library does not support fine-grained selective instrumentation, "+
				"which prevents safe auto-detection when apps partially instrument LangChain components. "+
				"Use langchain: true (platform instruments everything) or langchain: false (if your app instruments any LangChain components, even partially)",
		)
	}
	return nil
}

File: core/runtime-coordinator/agent_obs_runtime/plugins/langchain.py
"""LangChain instrumentation plugin.

LangChain does NOT support auto-detection because the official OTel instrumentor
(opentelemetry-instrumentation-langchain) uses an all-or-nothing approach that
instruments the entire LangChain ecosystem at once. This makes it unsafe to use
auto-detection when apps may partially instrument LangChain components.
"""

from __future__ import annotations

import logging

from .base import InstrumentationPlugin

LOGGER = logging.getLogger(__name__)


class LangChainPlugin(InstrumentationPlugin):
    """LangChain instrumentation plugin (no auto-detection support)."""

    @property
    def name(self) -> str:
        return "langchain"

    @property
    def supports_auto_detection(self) -> bool:
        # LangChain uses all-or-nothing instrumentation, not safe for auto-detection
        return False

    def should_instrument(self, config_value) -> bool:
        """Platform instruments only if explicitly configured as true."""
        return config_value is True

    def dependencies(self) -> list[str]:
        """Return LangChain instrumentation dependencies."""
        return ["opentelemetry-instrumentation-langchain>=0.51b0,<1.0"]

    def instrument(self):
        """Instrument LangChain using official OTel instrumentor."""
        try:
            from opentelemetry.instrumentation.langchain import LangchainInstrumentor

            LangchainInstrumentor().instrument()
            LOGGER.info("Instrumented LangChain")
        except Exception as exc:
            LOGGER.warning(f"Failed to instrument LangChain: {exc}")
            raise
